juju/juju
# Feature/ingress resources

## Please provide the following details to expedite Pull Request review:

### Checklist

 - [ ] <del>Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?</del>
 - [ ] <del>Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed? </del>
 - [ ] <del>Do comments answer the question of why design decisions were made? </del>

----

## Description of change

Introduce `ingress` resources in PodSpec `V2`;

## QA steps

- deploy CaaS charm with Ingress format like below;

```yaml
kubernetesResources:
  ingresses:
    - name: test-ingress
      labels:
        foo: bar
      annotations:
        ingress.kubernetes.io/rewrite-target: ""
        kubernetes.io/ingress.class: nginx
        kubernetes.io/ingress.allow-http: "true"
      spec:
        rules:
          - http:
              paths:
                - path: /
                  backend:
                    serviceName: test
                    servicePort: 80
```

- check result;

```console
$ kubectl get ingress.extensions/test-ingress -n <namespace> -o yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    ingress.kubernetes.io/rewrite-target: ""
    juju.io/controller: e34f5530-5850-4f32-8123-c36543aed23e
    juju.io/model: d6d8e529-dd1b-48b1-86aa-032185031702
    kubernetes.io/ingress.allow-http: "true"
    kubernetes.io/ingress.class: nginx
  creationTimestamp: "2019-11-27T07:03:58Z"
  generation: 1
  labels:
    foo: bar
    juju-app: mariadb-k8s
  name: test-ingress
  namespace: t1
  resourceVersion: "311230"
  selfLink: /apis/extensions/v1beta1/namespaces/t1/ingresses/test-ingress
  uid: 67983c83-f20f-4414-8722-948e9ef69ca7
spec:
  rules:
  - http:
      paths:
      - backend:
          serviceName: test
          servicePort: 80
        path: /
status:
  loadBalancer:
    ingress:
    - ip: 127.0.0.1
```


## Documentation changes

Yes

## Bug reference

No
